primitives/src/predicate.rs

```diff
@@ -0,0 +1,333 @@
+where
+    P: Predicate + Sized,
+{
+    NotPredicate::new(predicate)
+}
+
+/// Helper function to run predicates from a context.
+pub fn run(rule: Rule, context: &Context) -> bool {
+    match rule.rule_type {
+        RuleType::IsPresent(claim_type) => exists(claim_type).evaluate(context),
+        RuleType::IsAbsent(claim_type) => not(exists(claim_type)).evaluate(context),
+        RuleType::IsAnyOf(claims) => any(claims).evaluate(context),
+    }
+}
+
+// ExistentialPredicate
+// ======================================================
+
+/// It checks the existential of a claim.
+#[derive(Encode, Decode, Clone, Debug)]
+pub struct ExistentialPredicate {
+    /// Claims we want to check if it exists in context.
+    pub claim_type: ClaimType,
+}
+
+impl Predicate for ExistentialPredicate {
+    #[inline]
+    fn evaluate(&self, context: &Context) -> bool {
+        context
+            .claims
+            .iter()
+            .any(|ctx_claim| ctx_claim.claim_type() == self.claim_type)
+    }
+}
+
+// AndPredicate
+// ======================================================
+
+/// A composition predicate of two others using logical AND operator.
+#[derive(Encode, Decode, Clone, Debug)]
+pub struct AndPredicate<P1, P2>
+where
+    P1: Predicate,
+    P2: Predicate,
+{
+    lhs: P1,
+    rhs: P2,
+}
+
+impl<P1, P2> AndPredicate<P1, P2>
+where
+    P1: Predicate,
+    P2: Predicate,
+{
+    /// Create a new `AndPredicate` over predicates `lhs` and `rhs`.
+    #[inline]
+    pub fn new(lhs: P1, rhs: P2) -> Self {
+        AndPredicate { lhs, rhs }
+    }
+}
+
+impl<P1, P2> Predicate for AndPredicate<P1, P2>
+where
+    P1: Predicate,
+    P2: Predicate,
+{
+    /// Evaluate predicate against `context`.
+    #[inline]
+    fn evaluate(&self, context: &Context) -> bool {
+        self.lhs.evaluate(context) && self.rhs.evaluate(context)
+    }
+}
+
+// OrPredicate
+// ======================================================
+
+/// A composition predicate of two others using logical OR operator.
+#[derive(Encode, Decode, Clone, Debug)]
+pub struct OrPredicate<P1, P2>
+where
+    P1: Predicate,
+    P2: Predicate,
+{
+    lhs: P1,
+    rhs: P2,
+}
+
+impl<P1, P2> OrPredicate<P1, P2>
+where
+    P1: Predicate,
+    P2: Predicate,
+{
+    /// Create a new `OrPredicate` over predicates `lhs` and `rhs`.
+    #[inline]
+    pub fn new(lhs: P1, rhs: P2) -> Self {
+        OrPredicate { lhs, rhs }
+    }
+}
+
+impl<P1, P2> Predicate for OrPredicate<P1, P2>
+where
+    P1: Predicate,
+    P2: Predicate,
+{
+    /// Evaluate predicate against `context`.
+    #[inline]
+    fn evaluate(&self, context: &Context) -> bool {
+        self.lhs.evaluate(context) || self.rhs.evaluate(context)
+    }
+}
+
+// NotPredicate
+// ======================================================
+
+/// Predicate that returns a logical NOT of other predicate.
+#[derive(Encode, Decode, Clone, Debug)]
+pub struct NotPredicate<P: Predicate + Sized> {
```

Comment:
It's a side-effect because 'AndPredicate' & 'OrPredicate' are built from functions 'and' & 'or', whose have the `Size` constraint.
I am going to add this type constraint.